universe/auto_syncer.go

```diff
@@ -313,6 +452,51 @@ func (f *FederationEnvoy) syncer() {
 				continue
 			}
 
+			// After we've synced with the federation, we'll
+			// attempt to push out any pending proofs that we
+			// haven't yet completed.
+			ctxFetchLog, cancelFetchLog := f.WithCtxQuitNoTimeout()
+			syncDirection := SyncDirectionPush
+			logEntries, err :=
+				f.cfg.FederationDB.FetchPendingProofsSyncLog(
+					ctxFetchLog, &syncDirection,
+				)
+			if err != nil {
+				log.Warnf("unable to query pending push "+
+					"sync log: %w", err)
+				continue
+			}
+			cancelFetchLog()
+
+			// TODO(ffranr): Take account of any new servers that
+			//  have been added since the last time we populated the
+			//  log for a given proof leaf. Pending proof sync log
+			//  entries are only relevant for the set of servers
+			//  that existed at the time the log entry was created.
+			//  If a new server is added, then we should create a
+			//  new log entry for the new server.
+
+			for idx := range logEntries {
+				entry := logEntries[idx]
+
+				go func() {
```

Comment:
Aren't we already in a goroutine here? So can't we just wait for these operations to finish synchronously?

Reply:
If at this point we had a lot of pending log entries that we needed to handle, blocking at this point would stop the federation envoy from accepting new push requests. I think that's the only motivator I can think of. What do you think?

Reply:
Right, we would block new requests from coming in. But we aren't just logging something, we are pushing proofs to the universe. So we could run into a situation where we have a dependency or timing problem if something we rely on the universe knowing is not yet pushed because it's in a goroutine.
Also, `pushProofToFederation()` itself creates more goroutines, so we're spawning a lot of them potentially... I'd vote for blocking here instead.

Reply:
I agree with you that `pushProofToFederation` creates new go routines using `fn.ParSlice`. So I agree that we should remove the goroutine in the `for idx := range logEntries` loop.

> So we could run into a situation where we have a dependency or timing problem if something we rely on the universe knowing is not yet pushed because it's in a goroutine.

But at this point we're handling pending log entries kicked off by the federation envoy ticker. We've already waited until the next tick. We aren't populating proofs in the local universe we're pushing them to remote servers.

Reply:
Okay, I see that the timing concern probably isn't an issue. But still, creating goroutines in a loop is usually a red flag. And I guess in this instance, if there is reason to do it asynchronously, there should only be a single goroutine that does the whole loop. Not a goroutine per loop entry.

Reply:
> But still, creating goroutines in a loop is usually a red flag.

Ok, I'll keep that in mind :+1: